src/cargo/core/package.rs

```diff
@@ -518,16 +519,34 @@ impl<'gctx> PackageSet<'gctx> {
                 target_data,
                 force_all_targets,
             );
-            for (pkg_id, _dep) in filtered_deps {
+            for (pkg_id, deps) in filtered_deps {
                 collect_used_deps(
                     used,
                     resolve,
                     pkg_id,
                     has_dev_units,
-                    requested_kinds,
+                    requested_kind,
                     target_data,
                     force_all_targets,
                 )?;
+                let artifact_kinds = deps.iter().filter_map(|dep| {
```

Comment:
Can the rest of the changes in this commit be split out into a refactor commit?

Reply:
Not totally sure what you mean here. Are you talking about everything changed about collect_deps besides filtering for artifact kinds? The separate commit wouldn't compile.